### ISSUE TITLE
Should Condition operations inherit their parents dependencies?

### ISSUE THREAD

**Issue by anthonyherron** [danthorpe/Operations#147]
Hey Dan, I have a question somewhat related to [#129](https://github.com/danthorpe/Operations/issues/129). There are definitely ways around to get this to work, so perhaps it's a bit philosophical. I have a code example below:

```swift
		// Example condition that a login is required to happen
		class LoggedInCondition: OperationCondition {
			
			var name: String { return "\(self.dynamicType)" }
			var isMutuallyExclusive: Bool { return false }

			func dependencyForOperation(operation: Operation) -> NSOperation? {
				print("\(self.dynamicType): " + __FUNCTION__)
				return LoginOperation()
			}
			
			func evaluateForOperation(operation: Operation, completion: OperationConditionResult -> Void) {
				print("\(self.dynamicType): " + __FUNCTION__)
				completion(.Satisfied)
			}
		}
		
		// Let's imagine this class is used to perform some initial config
		class InitialOperation: Operation {
			private override func execute() {
				print("\(self.dynamicType): " + __FUNCTION__)
				// Arbitrary delay used to show order of execution
				let delayTime = dispatch_time(DISPATCH_TIME_NOW, Int64(5 * Double(NSEC_PER_SEC)))
				dispatch_after(delayTime, dispatch_get_main_queue()) {
					self.finish()
				}
			}
		}
		
		class LoginOperation: Operation {
			private override func execute() {
				print("\(self.dynamicType): " + __FUNCTION__)
				finish()
			}
		}
		
		class SimpleOperation: Operation {
			private override func execute() {
				print("\(self.dynamicType): " + __FUNCTION__)
				finish()
			}
		}
		
		let initialOperation = InitialOperation()
		let simpleOperation = SimpleOperation()
		
		simpleOperation.addDependency(initialOperation)
		// Before the simple operation can be performed, there's a requirement to be logged in
		simpleOperation.addCondition(LoggedInCondition())
		// Add the operations, would behave the same if this was a group
		operationQueue.addOperation(initialOperation)
		operationQueue.addOperation(simpleOperation)
```

Console output:
```
(LoggedInCondition #1): dependencyForOperation
(LoginOperation #1): execute()
(InitialOperation #1): execute()

// The delay happens here

(LoggedInCondition #1): evaluateForOperation(_:completion:)
(SimpleOperation #1): execute()
```

My question being... should the operation LoginOperation - which is returned by LoggedInCondition inherit the dependencies of SimpleOperation? Right now the operation is returned and executed concurrently. 

The reason I ask being the condition's operations may itself have InitialOperation as a dependency - we can get around that of course. But what about the fact that LoggedInCondition's operation is potentially needlessly executed if the InitialOperation's dependencies fail?

**Comment by anthonyherron** [comment]
I was going to write another separate ticket, but it's similar so I'll ask here: what about evaluating a condition before the conditions operation is added to the queue? So evaluateForOperation() is called, if it's already satisfied, then don't bother adding the condition to the queue. I don't know whether this could cause issues with prior operations, so I don't know whether a more "Just In Time" evaluation scenario would be more suitable. So they are evaluated in order after the parents dependencies are complete. This is what I did to the sample code form Apple, which worked for me, but obviously this suited my needs and wondered if there were any cases I haven't considered.

**Comment by danthorpe** [comment]
Hi @anthonyherron - thanks for the feedback & observations.

Firstly, I guess I've got to apologize for the lack of some clear documentation on this regarding the scheduling. I think that is mostly to blame for these questions because the exact order of operations regarding dependencies and conditions could be more clearly defined. On this, I am working on some more complete documentation.

But I can see sort of see why you've raised this issue. Essentially: If I have an operation with dependencies, shouldn't they be executed before any operations of any conditions? (Have I got that correct?)

And I guess I'm inclined to agree.

I would fix this in the queue: before adding the operations, we'd create another "finishing operation" which depends on all of the normal dependencies, sort of like a group. Then all of the operations from conditions would depend on that.

However my concerns are two fold.

1. Adding strict dependency rules like this unilaterally (where it will not always matter) reduces the multi-threaded nature of execution. The queue will start to be behave more serially which will impact performance. And it's hard to estimate the impact of this from my point of view (I've got no idea what kind of applications use the framework).

2. It's very hard to estimate how the change will impact existing codebases which use the repository. Yes it is definitely a breaking change - but might have to wait until the documentation is more evolved - so that the change can be better documented.

So, I'm sort of inclined to agree with you, and will make the changes. But they will be in conjunction with #129, and not until v2.6 at the earliest.

Also, I'd be interested to hear how @jshier uses this, as I believe he has a similar `LoggedInCondition` type behavior. Having some good examples showing best practices for these type of scenarios is a goal for the documentation. 

Regarding your second point... I have also wondered about making this more formal. Part of the problem is that conditions are evaluated asynchronously, but dependencies from conditions are returned synchronously. So checking whether the conditions are evaluated first would *really* complicate things in `Operation`. I think at the moment, if you have conditions which can be evaluated synchronously first - then inside `dependencyForOperation`, if it makes sense to do so - check if it's needed. You can see an example of this in [`AddressBookCondition`](https://github.com/danthorpe/Operations/blob/development/Sources/Extras/AddressBook/iOS/AddressBookConditions.swift). So, it's something which again, I would say should be a best practice for author's of `OperaitonCondition` types.

**Comment by anthonyherron** [comment]
Yes that's what I meant, dependent operations should be executed before condition operations. 

If you have a roadmap/TODO list, I'd be happy to contribute btw.

**Comment by danthorpe** [comment]
Hi @anthonyherron - I don't yet have a roadmap or contributing guidelines yet - but after I sort out 2.5 I'm going to put these in place. PRs are welcome - but things which drastically change how operations are executed need to be clearly understood / documented / assessed etc. And this is what I'm working on for  v2.5 which I hope will be done in the next day or so.

**Comment by danthorpe** [comment]
@anthonyherron I've implemented support for this in #157. If you have any feedback on it that would be great. Still got more specialized tests for this to write before it can be merged, but I think the scheduling on the queue is worth checking out regardless.

**Comment by danthorpe** [comment]
#157 has been merged now.

**Comment by danthorpe** [comment]
Doing some housekeeping - closing this issue now.

**Comment by anthonyherron** [comment]
Sorry for the belated reply @danthorpe - thank you for the fix, I'll take a look.

**Comment by danthorpe** [comment]
No worries @anthonyherron - I've had a bit of spare time so, burning through some issues. 

I think the info in #157 covered this issue quite well. But, bear in mind if you're looking at `development` that the changes proposed in #129 have also been done.

So, that diagram in #157 doesn't show that if there are mutually exclusive conditions which have dependencies, they will no longer be executed concurrently.
